### PR TITLE
Error resolutions

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Introduce error resolutions, a way to suppress errors by explaining why they can be ignored.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/951)

### CODE REVIEW

File: model/src/main/kotlin/config/ErrorResolutionReason.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * The error can not be fixed as it requires a fix to be made by 3rd party.
+     */
+    CANT_FIX_ISSUE,
+
+    /**
+     * The error is due to an irrelevant scanner issue, such as time out on a large file that is not distributed.
+     */
+    SCANNER_ISSUE
```
Comment (sschuberth) [inline model/src/main/kotlin/config/ErrorResolutionReason.kt]: This raises some questions for me: Should we have separate resolution reasons for the different steps, i.e. analyzer and scanner? I guess not, as it's more or less an implementation detail that ORT has two separate libraries for these steps. That said, maybe also some analyzer implementation could run into a timeout, or into another issue where a file could not be read completely. Should be combine these cases into a more general reason?

File: model/src/main/kotlin/config/ErrorResolutionReason.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+enum class ErrorResolutionReason {
+    /**
+     * The error is due to the way a package is built.
+     */
+    BUILD_TOOL_OF,
+
+    /**
+     * The error can not be fixed as it requires a fix to be made by 3rd party.
+     */
+    CANT_FIX_ISSUE,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/ErrorResolutionReason.kt]: Maybe: `NOT_FIXABLE` or `NEEDS_THIRD_PARTY_CHANGE`?

File: model/src/main/kotlin/config/ErrorResolutionReason.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package com.here.ort.model.config
+
+enum class ErrorResolutionReason {
+    /**
+     * The error is due to the way a package is built.
+     */
+    BUILD_TOOL_OF,
+
+    /**
+     * The error can not be fixed as it requires a fix to be made by 3rd party.
```
Comment (sschuberth) [inline model/src/main/kotlin/config/ErrorResolutionReason.kt]: I'd reword this a bit and say "The error can not be fixed, e.g. because it requires a change to be made by third party that is not responsive."

File: model/src/main/kotlin/config/ErrorResolutionReason.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.model.config
+
+enum class ErrorResolutionReason {
+    /**
+     * The error is due to the way a package is built.
+     */
+    BUILD_TOOL_OF,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/ErrorResolutionReason.kt]: We have the same enum member in `ProjectExcludeReason` and `ScopeExcludeReason` already, which maybe creates confusion when to use which. What's an example for this?

File: model/src/main/kotlin/config/ErrorResolution.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+package com.here.ort.model.config
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+
+/**
+ * Defines the resolution of an error. This can be used to silence false positives, or errors that have been identified
+ * as not being relevant.
+ */
+data class ErrorResolution(
```
Comment (sschuberth) [inline model/src/main/kotlin/config/ErrorResolution.kt]: Should we more generally call this `IssueResolution` already now, so we could also use it to resolve warnings (once we add them) later?

File: model/src/main/kotlin/config/Resolutions.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Resolutions for issues with a repository.
+ */
+class Resolutions(
+    /**
+     * Resolutions for issues with the analysis or scan of the projects in this repository and their dependencies.
+     */
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    val errors: List<ErrorResolution> = emptyList()
```
Comment (sschuberth) [inline model/src/main/kotlin/config/Resolutions.kt]: Note to myself: I was first wondering why this is called "errors" instead of "resolutions" (because they are more the latter than the former), but currently it reads nicer in the serialized output, and we could add a "warnings" field later, too.

File: reporter/src/main/kotlin/DefaultResolutionProvider.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     * A [Resolutions] object, typically taken from a [RepositoryConfiguration].
+     */
+    private val repositoryResolutions: Resolutions,
+
+    /**
+     * A JSON or YAML encoded file containing resolutions.
+     */
+    private val resolutionsFile: File? = null
+) : ResolutionProvider {
+    private val fileResolutions: Resolutions  by lazy { resolutionsFile?.readValue() ?: Resolutions() }
```
Comment (sschuberth) [inline reporter/src/main/kotlin/DefaultResolutionProvider.kt]: Nit: Double space here.

Comment (sschuberth) [reply reporter/src/main/kotlin/DefaultResolutionProvider.kt]: The variable name sounds like resolutions *for* files. Maybe `globalResolutions` (or `customResolutions` or `generalResolutions`, and `repositoryResolutions`)?

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 8]
```diff
@@ -82,9 +92,12 @@ object ReporterCommand : CommandWithHelp() {
 
         val ortResult = ortResultFile.readValue<OrtResult>()
 
+        val resolutionProvider =
+            DefaultResolutionProvider(ortResult.repository.config.resolutions ?: Resolutions(), resolutionsFile)
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ReporterCommand.kt]: Here it occurs to me that we should probably have a `ResolutionFileProvider` and a list of `ResolutionProviders` (with two instanced of `ResolutionFileProvider`) instead of passing two files to a `DefaultResolutionProvider`.

Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/ReporterCommand.kt]: Working on it.

Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/ReporterCommand.kt]: Took a different approach, see `DefaultResolutionProvider` in the updated code.

File: reporter/src/main/kotlin/DefaultResolutionProvider.kt (hunk 7 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/DefaultResolutionProvider.kt]: Removed the variables in the latest iteration of the PR.

File: model/src/main/kotlin/config/ErrorResolution.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * The reason why the errors is resolved.
+     */
+    val reason: ErrorResolutionReason,
+
+    /**
+     * A comment to further explain why the [reason] is applicable here.
+     */
+    val comment: String
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply model/src/main/kotlin/config/ErrorResolution.kt]: Well, maybe we could use [sealed classes](https://kotlinlang.org/docs/reference/sealed-classes.html) instead?

File: model/src/main/kotlin/config/ErrorResolutionReason.kt (hunk 1 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/ErrorResolutionReason.kt]: Yes, I like the latter somewhat.

**Comment by sschuberth** [comment]
Commit message nits (commits are identified by their title, nits mentioned below it):

> model: Add model classes for error resolutions

"Error resolution*s*"

> model: Add a resolutions property to `RepositoryConfiguration`

"A resolution" (because you refer as "it" to it later.)

> reporter: Add a `ResolutionProvider` interface to access resolutions

Need to drop the body text about `DefaultResolutionProvider` now.

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 10]
```diff
@@ -29,17 +29,23 @@ import com.beust.jcommander.Parameters
...
 import com.here.ort.utils.showStackTrace
 
 import java.io.File
 
-@Parameters(commandNames = ["report"],
-        commandDescription = "Present Analyzer and Scanner results in various formats.")
+@Parameters(
+    commandNames = ["report"],
+    commandDescription = "Present Analyzer and Scanner results in various formats."
+)
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ReporterCommand.kt]: Is this reformatting (and below) intended?

File: reporter/src/main/kotlin/DefaultResolutionProvider.kt [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.reporter
+
+import com.here.ort.model.Error
+import com.here.ort.model.config.Resolutions
+
+class DefaultResolutionProvider : ResolutionProvider {
```
Comment (sschuberth) [inline reporter/src/main/kotlin/DefaultResolutionProvider.kt]: Should this maybe more explicitly be called `ResolutionProviderCollection` or so, to describe its purpose in the name somewhat?

File: model/src/main/kotlin/config/ErrorResolution.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply model/src/main/kotlin/config/ErrorResolution.kt]: Probably, but Jackson deserialization with inheritance can be tricky, and could require to add type information on serialization. Also, as resolutions are very similar to excludes I wouldn't want to implement them in a different way.

File: model/src/main/kotlin/config/ErrorResolutionReason.kt (hunk 4 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/config/ErrorResolutionReason.kt]: I think it's good to reuse the same names, makes it easier to remember.
One example would be a Gradle project using a `flatDir` repository. This will generate an analyzer error because we do not support this, which could be resolved using `BUILD_TOOL_OF`.

Comment (sschuberth) [reply model/src/main/kotlin/config/ErrorResolutionReason.kt]: Then let's use `BUILD_TOOL_ISSUE` [as suggested by Thomas](https://github.com/heremaps/oss-review-toolkit/issues/731#issuecomment-412806641) but extend the docs to e.g. "The error originates from the build tool used by the project".

Comment (mnonnenmacher) [reply model/src/main/kotlin/config/ErrorResolutionReason.kt]: Done.

File: cli/src/main/kotlin/commands/ReporterCommand.kt (hunk 10 above)
Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/ReporterCommand.kt]: No, undone.

File: reporter/src/main/kotlin/DefaultResolutionProvider.kt (hunk 11 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/DefaultResolutionProvider.kt]: I find this other name also a bit confusing and couldn't think of a better one. Let's keep it for now.

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 12]
```diff
@@ -73,6 +76,11 @@ object ReporterCommand : CommandWithHelp() {
             order = PARAMETER_ORDER_MANDATORY)
     private lateinit var reportFormats: List<Reporter>
 
+    @Parameter(description = "A JSON or YAML file that contains resolutions.",
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ReporterCommand.kt]: We don't say which file formats we support in the other parameters, so I guess we shouldn't here either. But instead we should probably mention that this is about *error* resolutions, like "A file containing error resolutions".

Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/ReporterCommand.kt]: Done.

**Comment by mnonnenmacher** [comment]
Merging, ignoring the unrelated AppVeyor error.
